### ISSUE TITLE
Consider Including Cancellable Combinator

### ISSUE THREAD

**Issue by cramertj** [rust-lang-nursery/futures-rs#693]
I find myself wanting this occasionally. [Example here](https://play.rust-lang.org/?gist=d49b79a00aba4f935ff42dfe5ef5c65e&version=stable).

**Comment by alexcrichton** [comment]
Seems nifty! It sort of seems like you're doing something akin to:

```rust
let future = ...;
let (tx, rx) = sync::oneshot::channel();
let cancellable_future = rx.select(future);
let cancellable_handel = || drop(tx.send(());
```

Although the implementation here definitely seems more optimized than that is. The name "cancellable" though may not be best here in the sense that all futures are inherently cancellable (in that you just drop them) but something like a "cancel handle" makes sense to me. Something like "kill switch" may be a bit too egregious though!

**Comment by cramertj** [comment]
Yeah, that's the idea-- this is a bit more optimized and I think the intent is clearer. If you're interested, I'll open a PR and we can bikeshed on the name there.

**Comment by alexcrichton** [comment]
Seems plausible to me!

**Comment by tikue** [comment]
I don't know the inner workings of `oneshot::channel` -- I'm curious, how is this more optimized than selecting over a oneshot receiver?

**Comment by taiki-e** [comment]
@cramertj I think you implemented this as `Abortable` in #1079.

**Comment by cramertj** [comment]
Indeed! Will close.
